oap-server/server-receiver-plugin/envoy-metrics-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/envoy/AccessLogServiceGRPCHandler.java

```diff
@@ -105,7 +110,7 @@ public AccessLogServiceGRPCHandler(ModuleManager manager,
             public void onNext(StreamAccessLogsMessage message) {
                 HistogramMetrics.Timer timer = histogram.createTimer();
                 try {
-                    if (isFirst) {
+                    if (isFirst || (alwaysAnalyzeIdentity && message.getIdentifier() != null)) {
```

Comment:
Reply to https://github.com/apache/skywalking/pull/8193#issuecomment-980809798

@kezhenxu94 I don't think this will affect the parsing here. The parsing is already done, this if protobuf, not flatbuffer.
So, the only one more cost is reading `message.getIdentifier()`, which is a Java bean getting only.

Reply:
You overlooked these:

```java
            for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                 role = analysis.identify(identifier, role);
            }
```

Reply:
How about this? Logic inside has nothing relating to this added flag. I am not following your question.
No matter in which case, this `identifier ` is required. The difference is from `StreamObserver#cache` or next element.

Reply:
The difference to me is 

# old

```java
if (isFirst) { // only true for the first element per stream
            for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                 role = analysis.identify(identifier, role);
            }
}
```

# new

```java
if (true) { // it is no 100% always true but is extremely high possibility true from my understanding
            for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                 role = analysis.identify(identifier, role);
            }
}
```

does this illustrate my concern? @wu-sheng 

If @mrproliu you had experimented this, can you just post the so11y metrics (especially CPU) of OAP before and after the new protocol?

Reply:
If you mean flatbuffer parsing, them this is really having no different than before.
I think your gap is that, you are not considering we merged the stream, currently, one stream is representing the previous multiple streaming, then logically, all those identifiers are pasred too, just through different streamings.

Reply:
> If @mrproliu you had experimented this, can you just post the so11y metrics (especially CPU) of OAP before and after the new protocol?

I could not happen on this, because it decides by the satellite message package mechanism. Only the first message in the batch could have the identity.

Reply:
I am not sure why this possiblily could be high. You are comparing this logic between one stream and 2k+ streaming. Then you have a conclusion, this one streaming has higher possible to parse identify.
More parsing per stream, yes, but no different when condidering this whole 2k streamings(2k envoy instances)

Reply:
If you want I could do the experiment, we also using the new protocol and the `OAP` side is using current and this one. Then I post the so11y metrics.
```java```
if (isFirst || (alwaysAnalyzeIdentity && message.getIdentifier() != null)) {
                        identifier = message.getIdentifier();
                        isFirst = false;
                        role = Role.NONE;
                    }

                    for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                        role = analysis.identify(identifier, role);
                    }
```java```
I think only this one could find the identity analysis whether there is a performance problem.

Reply:
> If you mean flatbuffer parsing, then this is really having no different than before.

I never mean flatbuffer parsing. I meant this part 

```java
for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                 role = analysis.identify(identifier, role);
}
```

> I think your gap is that, you are not considering we merged the stream, currently, one stream is representing the previous multiple streaming, then logically, all those identifiers are pasred too, just through different streamings.

I’m rather aware of that you merged the streams, and because you merge the streams from different Envoy nodes the OAP needs to identify the node in every message

Reply:
@wu-sheng forget the word “parse”, I don’t mean parsing flatbuffer, I mean identifying the role of the node 

```java
for (ALSHTTPAnalysis analysis : envoyHTTPAnalysisList) {
                 role = analysis.identify(identifier, role);
            }
```

Reply:
Let’s have a DM discuss if you still don’t get my point

Reply:
I think I know what is the gap here. Let me put in this way

1. `Envoy -> OAP` direct linking

All envoy open unclosed streaming because the connection is always open, then the identifier only gets parsed once. This costs minimal in identifier parsing, but we have been facing load not balanced. Then we were introducing Satellite.

2. `Envoy->Satellite->OAP`, current satellite implementation.

It maintains the context for a bulk of ALS message(such as 100 per bulk), then it created stream every time. With this solution, the load is balanced like you have seen in previous Satellite PR. 
This change doesn't require any OAP side change, but actually, it already introduce the extra identifier parsing logic, you raised here(which is why I can't understand for a while, this is never introduced by this change).
One side, maintaining unclosed streaming in Satellite is impossible(costing more CPUs), also, keeping open/close streaming(current status) is also not good.

Then, we get the idea of this new protocol and new PR.

We are merging these streamings as one new streaming(unclosed this time) but carrying different identifiers.

Reply:
What you are concerned and asked about actually is already there after local balancer added. So, generally, you need a tradeoff, because trying to balance 2k+ envoy instances, the streaming must be merged one way or another.

But what is cost could be controlled by what is the size of Satellite side bulk setting. Higher is set, less CPU(parsing identifier) used, but more memory is used at Satellite.

Reply:
So, generally, your asking about more resource cost at OAP side is true, but not because of this flag, but load balancing mechanism.

Reply:
> What you are concerned and asked about actually is already there after local balancer added. So, generally, you need a tradeoff, because trying to balance 2k+ envoy instances, the streaming must be merged one way or another.

I agree we need tradeoff and I've also given another merging strategy for discussion, which I think might be better https://github.com/apache/skywalking/pull/8193#issuecomment-980766033

> So, generally, your asking about more resource cost at OAP side is true, but not because of this flag, but load balancing mechanism.

I doubt this and need to confirm with @mrproliu .

Despite of all above, `if (isFirst || (alwaysAnalyzeIdentity && message.getIdentifier() != null))` is buggy because `message.getIdentifier()` is never null, it returns a default instance if `identifier == null`

Reply:
>   // Identifier data that will only be sent in the first message on the stream. This is effectively
  // structured metadata and is a performance optimization.
  Identifier identifier = 1;

This is from ALS definition. What do you mean `never null?

Reply:
I said this from code level. `message.getIdentifier()` never returns null, protocol buffer compiler generates codes like `return identifier == null ? defaultInstance : identifier`, please use `message.hasIdentifier()`

Reply:
OK, make sense.

Reply:
Fixed. thanks.

Reply:
I can't see the commit.